subsys/net/ip/ipv4_autoconf.c

```diff
@@ -0,0 +1,280 @@
+
+	net_ipaddr_copy(&hdr->dst_ipaddr, &iface->ipv4.current_ip);
+	net_ipaddr_copy(&hdr->src_ipaddr, &iface->ipv4.requested_ip);
+
+	net_buf_add(frag, sizeof(struct net_arp_hdr));
+
+	return pkt;
+
+fail:
+	net_pkt_unref(pkt);
+	return NULL;
+}
+
+enum net_verdict net_ipv4_autoconf_input(struct net_if *iface,
+				struct net_pkt *pkt)
+{
+	struct net_arp_hdr *arp_hdr;
+
+	if (net_pkt_get_len(pkt) < (sizeof(struct net_arp_hdr) -
+				    net_pkt_ll_reserve(pkt))) {
+		NET_DBG("Invalid ARP header (len %zu, min %zu bytes)",
+			net_pkt_get_len(pkt),
+			sizeof(struct net_arp_hdr) -
+			net_pkt_ll_reserve(pkt));
+		return NET_DROP;
+	}
+
+	arp_hdr = NET_ARP_HDR(pkt);
+
+	if (!net_ipv4_addr_cmp(&arp_hdr->dst_ipaddr,
+			       &iface->ipv4.requested_ip)) {
+		/* No conflict */
+		return NET_DROP;
+	}
+
+#if NET_DEBUG > 0
+	do {
+		NET_DBG("conflict detected from %s [%s] for %s", out,
+			net_sprint_ll_addr((uint8_t *)&arp_hdr->src_hwaddr,
+					   arp_hdr->hwlen),
+			net_sprint_ipv4_addr(&arp_hdr->dst_ipaddr));
+	} while (0);
+#endif
+
+	iface->ipv4.conflict_cnt++;
+
+	switch (iface->ipv4.state) {
+	case NET_IPV4_AUTOCONF_PROBE:
+		/* restart probing with renewed IP */
+		net_ipv4_autoconf_start(iface);
+		break;
+	case NET_IPV4_AUTOCONF_ANNOUNCE:
+	case NET_IPV4_AUTOCONF_CONFIGURED:
+		if (iface->ipv4.conflict_cnt == 1) {
+			/* defend IP */
+			send_announcement(iface);
+		} else {
+			/* unset host ip */
+			if (!net_if_ipv4_addr_rm(
+				    iface, &iface->ipv4.requested_ip)) {
+				NET_DBG("Failed to remove addr from iface");
+			}
+			/* restart probing after second conflict */
+			net_ipv4_autoconf_start(iface);
+		}
+
+		break;
+	default:
+		break;
+	}
+
+	return NET_DROP;
+}
+
+static void ipv4_autoconf_timeout(struct k_work *work)
+{
+	struct net_if *iface =
+		CONTAINER_OF(work, struct net_if, ipv4.ipv4_autoconf_timeout);
+
+	if (!iface) {
+		NET_DBG("Invalid iface");
+		return;
+	}
+
+	switch (iface->ipv4.state) {
+	case NET_IPV4_AUTOCONF_PROBE:
+		/* schedule next probe */
+		if (iface->ipv4.probe_cnt <=
+		    IPV4_AUTOCONF_PROBE_NUM - 1) {
+			send_probe(iface);
+			break;
+		}
+
+	case NET_IPV4_AUTOCONF_ANNOUNCE:
+		if (iface->ipv4.announce_cnt <=
+		    IPV4_AUTOCONF_ANNOUNCE_NUM - 1) {
+			net_ipaddr_copy(&iface->ipv4.current_ip,
+					&iface->ipv4.requested_ip);
+			send_announcement(iface);
+			break;
+		}
+
+		/* success, add new IPv4 address */
+		if (!net_if_ipv4_addr_add(iface,
+					  &iface->ipv4.requested_ip,
+					  NET_ADDR_AUTOCONF, 0)) {
+			NET_DBG("Failed to add IPv4 addr to iface %p", iface);
+			return;
+		}
+
+		iface->ipv4.state = NET_IPV4_AUTOCONF_CONFIGURED;
+		break;
+	default:
+		break;
+	}
+}
+
+static inline uint32_t ipv4_autoconf_get_timeout(struct net_if *iface)
+{
+	switch (iface->ipv4.state) {
+	case NET_IPV4_AUTOCONF_PROBE:
+		if (iface->ipv4.conflict_cnt >=
+		    IPV4_AUTOCONF_MAX_CONFLICTS) {
+			NET_DBG("rate limiting\n");
+			return IPV4_AUTOCONF_RATE_LIMIT_INTERVAL *
```

Comment:
If IPV4_AUTOCONF_RATE_LIMIT_INTERVAL is in seconds, then just do
K_SECONDS(IPV4_AUTOCONF_RATE_LIMIT_INTERVAL)
here so there would be no need to multiply by MSEC_PER_SEC.